proxy/http/HttpTunnel.h

```diff
@@ -475,12 +475,28 @@ HttpTunnel::get_producer(HttpTunnelType_t type)
 inline HttpTunnelConsumer *
 HttpTunnel::get_consumer(VConnection *vc)
 {
-  for (int i = 0; i < MAX_CONSUMERS; i++) {
-    if (consumers[i].vc == vc) {
-      return consumers + i;
+  /** Rare but persistent problem in which a @c INKVConnInternal is used a consumer, released, and
+      then re-allocated for a different consumer. This causes two consumers to have the same VC
+      pointer resulting in this method returning the wrong consumer. Note this is a not a bad use of
+      the tunnel, but an unfortunate interaction with the FIFO free lists.
+
+      It's not correct to check for the consumer being alive - at a minimum `HTTP_TUNNEL_EVENT_DONE`
+      is dispatched against a consumer after the consumer is not alive. Instead if a non-alive
+      consumer matches it is stored as a candidate and returned if no other match is found. If a
+      live matching consumer is found, it is immediately returned. It is never valid to have the
+      same VC in more than one active consumer. This should avoid a performance impact because in
+      the usual case the consumer will be alive.
+  */
+  HttpTunnelConsumer *zret = nullptr;
+  for (HttpTunnelConsumer &c : consumers) {
+    if (c.vc == vc) {
```

Comment:
The "next" consumer and "the one that matches" are the same thing in almost all cases. In general every "vc" (which are `InkContInternal` instances) has a distinct address so there is only one match, which is by definition therefore the "next" consumer. In both the new and old code, the consumer vc is checked for a pointer match against the argument vc. The logic differs only in the case the same vc address is in the table for two consumers.

The old code worked fine as long as every `InkContInternal` allocated during the span of a transaction had a distinct address, which is quite normally the case. Although the free list is a FIFO, if any other transaction or thread allocated between a transaction's release and realloc there would not be a problem. You need just the right amount of load (some, but not too much) and the right timing on the transactions on the same `ET_NET` thread to trigger this.